src/Robot/TableauFoundation.hs

```diff
@@ -242,63 +257,71 @@ getPrefixDiff (a:as) (b:bs) = if a == b then getPrefixDiff as bs else Nothing
 -- which targs we're allowed to solve using the hyps from the given boxes)
 -- e.g. boxNumbersToDirections [([1,0,3], "bottom"),([1,0], "middle"),([], "root")]
 -- = Just ([([], "root"), ([1, 0], "middle"), ([3], "bottom")], [1,0,3])
-boxNumbersToDirections :: [(BoxNumber, a)] -> Maybe ([(BoxNumber, a)], BoxNumber)
+boxNumbersToDirections :: (MonadPlus m) => [(BoxNumber, a)] ->
+  m ([(BoxNumber, a)], BoxNumber)
 boxNumbersToDirections boxNumbersWithData = let
-  reverseSortedBoxNumbersWithData = sortBy (\b a -> length (fst a) `compare` length (fst b)) boxNumbersWithData
-  traverseBoxNumbers :: [(BoxNumber, a)] -> [(BoxNumber, a)] -> Maybe [(BoxNumber, a)]
-  traverseBoxNumbers trailSoFar [] = Just trailSoFar
-  traverseBoxNumbers trailSoFar [thisBox] = Just (thisBox:trailSoFar)
+  reverseSortedBoxNumbersWithData = sortBy
+    (\b a -> length (fst a) `compare` length (fst b)) boxNumbersWithData
+  traverseBoxNumbers :: (MonadPlus m) => [(BoxNumber, a)] -> [(BoxNumber, a)] ->
```

Comment:
This is pretty minor but it's a little strange to me because this is technically a function for all `m`, whereas what we really want is to be using the same `m` from earlier - the extension ScopedTypeVariables makes this work more nicely (and it's a relatively tiny extension). No need to change for now though